### PR TITLE
Add feature gate for StorageNamespaceIndex in v1.30

### PR DESCRIPTION
**Pull request by tengqm**

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  list: never
+  render: false
+
+stages:
+  - stage: beta
+    defaultValue: true
+    fromVersion: "1.30"
+---
+Allows namespace indexer for namespace scoped resources in API server cache
+to accelerate list operations.
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md:14]: I'm afraid I'd have to read the code to figure out what this means.

Comment (tengqm) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md:14]: Sadly, this is true. We have witnessed a tendency to feature gate abuse.
Whenever the developers are unsure whether a change is good, they add a new feature gate.
Sometimes the new "gate" is not about a user-facing feature. It is merely an
improvement to the existing behavior which may and may not be perceived by the users.
A gate is added anyway.
I'm really doubting whether there are people playing with this kind of gates.
I'm also concerning whether we as a community has a well established criteria for gates.
To an extreme end, we can add gate for whatever changes introduced. The reasons can be
that the requirements are unclear, the implementation is not solid, the effects are
yet to be evaluated, so on and so forth.
In an ideal world, feature gates should be constrained to changes that really make
a big difference for the users, changes that are understandable for users.
But, you know, the real life always sucks.

**Comment by drewhagen** [comment]
- Searched for [WatchListClient in Kubernetes code base](https://github.com/kubernetes/kubernetes/pull/122571/files#diff-b7315e8b207a5ae5ec05d07eb589aead09d071b9b29cf9e4b82d469b234bbb0cR39)
- Tracked to this [code PR 121906](https://github.com/kubernetes/kubernetes/pull/121906/files#diff-71e3b98f9a6bbf5b8421e26a7ba0c079f397cd8d49abacdad943c66a4f44f03dR983).
- In the comments, the decision was not to set up a KEP for this contribution, so there's none.
/sig api-machinery 

cc: @ahutsunshine @kubernetes/sig-api-machinery-misc

Can we please get some help contributing some detail to this feature gate doc? We appreciate it, thanks!

File: content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md (hunk 1 above)
Comment (ahutsunshine1) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md:14]: @sftim @tengqm For the namespace indexer feature, we did a long time discussion and did a benchmark that shown excellent performance in terms of latency, CPU, and memory in large clusters (like with 100k+ pods) for listing pods within a single namespace with namespace indexers. This implementation significantly reduces cpu and memory consumption, especially when dealing with a small number of pods.
Discussion:
- https://github.com/kubernetes/kubernetes/issues/120778
Benchmark doc: 
- https://docs.google.com/document/d/1bTxGqlH0tqrpH16NxC2SIUTGMN3_1zNKRiRdIW9Y5VU/edit?usp=sharing

Comment (ahutsunshine1) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md:14]: Also, we are from eBay Cloud team and this feature already is applied to our production that runs good performance.

**Comment by ahutsunshine1** [comment]
> * Searched for [`StorageNamespaceIndex` in Kubernetes code base](https://github.com/search?q=repo%3Akubernetes%2Fkubernetes%20StorageNamespaceIndex&type=code)
> * Tracked to this [code PR 121906](https://github.com/kubernetes/kubernetes/pull/121906/files#diff-71e3b98f9a6bbf5b8421e26a7ba0c079f397cd8d49abacdad943c66a4f44f03dR983).
> * In the comments, the decision was not to set up a KEP for this contribution, so there's none.
> 
> /sig api-machinery
> 
> cc: @ahutsunshine @kubernetes/sig-api-machinery-misc
> 
> Can we please get some help contributing some detail to this feature gate doc? We appreciate it, thanks!

I'm sorry for late response. My account has some issues, so I register https://github.com/ahutsunshine1 account to answer this question.  Previously we already did a long time discussion and did a benchmark for the namespace indexer feature that shown excellent performance in terms of latency, CPU, and memory in large clusters (like with 100k+ pods) for listing pods within a single namespace with namespace indexers. This implementation significantly reduces cpu and memory consumption, especially when dealing with a small number of pods.

Discussion:
- https://github.com/kubernetes/kubernetes/issues/120778

Benchmark doc:
- https://docs.google.com/document/d/1bTxGqlH0tqrpH16NxC2SIUTGMN3_1zNKRiRdIW9Y5VU/edit?usp=sharing

We, eBay Cloud team, already applied the namespace indexer feature to our production and run good performance.

**Comment by sftim** [comment]
Can you explain what enabling the feature gate means @ahutsunshine1 ?

You're explaining this to an audience that has never looked at one line of Kubernetes code, doesn't know Go syntax, and who aren't familiar with the performance challenges from not applying the fix.

**Comment by ahutsunshine1** [comment]
> Can you explain what enabling the feature gate means @ahutsunshine1 ?
> 
> You're explaining this to an audience that has never looked at one line of Kubernetes code, doesn't know Go syntax, and who aren't familiar with the performance challenges from not applying the fix.

@sftim After enabling the feature gate, it benefits the LIST namespace-scoped pods like `/api/v1/namespaces/default/pods?resourceVersion=0`.  Simply:

1. Before enabling, `/api/v1/namespaces/default/pods?resourceVersion=0` will search through all the pods from all namespaces in the cache and then filter the results based on the namespace `default`. For large clusters with more than 100k+ pods, it will spend lots of cpu and memory with the similar high LIST pods requests. 
2. After enabling, `/api/v1/namespaces/default/pods?resourceVersion=0` will get pods from `default` namespace indexer directly. 
If you want to know more, I would like to recommend to read the benchmark doc with the detail https://docs.google.com/document/d/1bTxGqlH0tqrpH16NxC2SIUTGMN3_1zNKRiRdIW9Y5VU/edit.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md (hunk 1 above)
Comment (ahutsunshine1) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/storage-namespace-index.md:14]: Yes, just pods at the beginning from the recommendation @wojtek-t . If stable after running for some time, I would like to enable this for the major resources(We already enabled major resources in our internal kubernetes controlplane).

**Pull request closed without merging**